src/client/streamableHttp.ts

```diff
@@ -0,0 +1,370 @@
+      }
+    }
+
+    if (this._sessionId) {
+      headers["mcp-session-id"] = this._sessionId;
+    }
+
+    return headers;
+  }
+
+  private async _startOrAuth(): Promise<void> {
+    try {
+      // Try to open an initial SSE stream with GET to listen for server messages
+      // This is optional according to the spec - server may not support it
+      const commonHeaders = await this._commonHeaders();
+      const headers = new Headers(commonHeaders);
+      headers.set('Accept', 'text/event-stream');
+
+      // Include Last-Event-ID header for resumable streams
+      if (this._lastEventId) {
+        headers.set('last-event-id', this._lastEventId);
+      }
+
+      const response = await fetch(this._url, {
+        method: 'GET',
+        headers,
+        signal: this._abortController?.signal,
+      });
+
+      if (response.status === 405 || response.status === 404) {
+        // Server doesn't support GET for SSE, which is allowed by the spec
+        // We'll rely on SSE responses to POST requests for communication
+        return;
+      }
+
+      if (!response.ok) {
+        if (response.status === 401 && this._authProvider) {
+          // Need to authenticate
+          return await this._authThenStart();
+        }
+
+        const error = new Error(`Failed to open SSE stream: ${response.status} ${response.statusText}`);
+        this.onerror?.(error);
+        throw error;
+      }
+
+      // Successful connection, handle the SSE stream as a standalone listener
+      const streamId = `initial-${Date.now()}`;
+      this._handleSseStream(response.body, streamId);
+    } catch (error) {
+      this.onerror?.(error as Error);
+      throw error;
+    }
+  }
+
+  async start() {
+    if (this._activeStreams.size > 0) {
+      throw new Error(
+        "StreamableHTTPClientTransport already started! If using Client class, note that connect() calls start() automatically.",
+      );
+    }
+
+    this._abortController = new AbortController();
+    return await this._startOrAuth();
+  }
+
+  /**
+   * Call this method after the user has finished authorizing via their user agent and is redirected back to the MCP client application. This will exchange the authorization code for an access token, enabling the next connection attempt to successfully auth.
+   */
+  async finishAuth(authorizationCode: string): Promise<void> {
+    if (!this._authProvider) {
+      throw new UnauthorizedError("No auth provider");
+    }
+
+    const result = await auth(this._authProvider, { serverUrl: this._url, authorizationCode });
+    if (result !== "AUTHORIZED") {
+      throw new UnauthorizedError("Failed to authorize");
+    }
+  }
+
+  async close(): Promise<void> {
+    // Close all active streams
+    for (const reader of this._activeStreams.values()) {
+      try {
+        reader.cancel();
+      } catch (error) {
+        this.onerror?.(error as Error);
+      }
+    }
+    this._activeStreams.clear();
```

Comment:
It might be easier to use a single `AbortController` for the whole object, then pass its `AbortSignal` into all the fetch/SSE calls.